src/awkward/_v2/operations/structure/ak_fill_none.py

```diff
@@ -2,125 +2,133 @@
+
+    #     The list content now has a union type:
+
+    #         >>> ak.type(ak.fill_none(array, "hi"))
+    #         3 * var * union[float64, string]
+
+    #     The values could be floating-point numbers or strings.
+    #     """
+
+    arraylayout = ak._v2.operations.convert.to_layout(
+        array, allow_record=True, allow_other=False
+    )
+    nplike = ak.nplike.of(arraylayout)
+
+    # Convert value type to appropriate layout
+    if (
+        isinstance(value, np.ndarray)
+        and issubclass(value.dtype.type, (np.bool_, np.number))
+        and len(value.shape) != 0
+    ):
+        valuelayout = ak._v2.operations.convert.to_layout(
+            nplike.asarray(value)[np.newaxis], allow_record=False, allow_other=False
+        )
+    elif isinstance(value, (bool, numbers.Number, np.bool_, np.number)) or (
+        isinstance(value, np.ndarray)
+        and issubclass(value.dtype.type, (np.bool_, np.number))
+    ):
+        valuelayout = ak._v2.operations.convert.to_layout(
+            nplike.asarray(value), allow_record=False, allow_other=False
+        )
+    elif (
+        isinstance(value, Iterable)
+        and not (
+            isinstance(value, (str, bytes))
+            or (ak._v2._util.py27 and isinstance(value, ak._v2._util.unicode))
+        )
+        or isinstance(value, (ak._v2.highlevel.Record, ak._v2.record.Record))
+    ):
+        valuelayout = ak._v2.operations.convert.to_layout(
+            value, allow_record=True, allow_other=False
+        )
+        if isinstance(valuelayout, ak._v2.record.Record):
+            valuelayout = valuelayout.array[valuelayout.at : valuelayout.at + 1]
+        elif len(valuelayout) == 0:
+            offsets = ak._v2.index.Index64(nplike.array([0, 0], dtype=np.int64))
+            valuelayout = ak._v2.contents.ListOffsetArray(offsets, valuelayout)
+        else:
+            valuelayout = ak._v2.contents.RegularArray(valuelayout, len(valuelayout), 1)
+    else:
+        valuelayout = ak._v2.operations.convert.to_layout(
+            [value], allow_record=False, allow_other=False
+        )
+
+    def maybe_fillna(layout):
+        if layout.is_OptionType:
+            return layout.fillna(valuelayout)
+        else:
+            return layout
+
+    if axis is None:
+
+        def transform(layout, depth, posaxis):
+            return ak._v2._util.transform_child_layouts(
+                transform, maybe_fillna(layout), depth, posaxis
+            )
+
+    else:
+
+        def transform(layout, depth, posaxis):
+            posaxis = layout.axis_wrap_if_negative(posaxis)
+            if posaxis + 1 < depth:
+                return layout
+
+            if posaxis + 1 == depth:
+                layout = maybe_fillna(layout)
+
+            return ak._v2._util.transform_child_layouts(
+                transform, layout, depth, posaxis
+            )
+
+    out = transform(arraylayout, 1, axis)
```

Comment:
Instead of these three calls to `transform`, use `recursively_apply`. (The intention was to reduce all these different ways of doing essentially the same thing with a single interface—that's why I don't want all the old interfaces brought back.)